examples/mustache_template_motd.cf

```diff
@@ -1,71 +1,120 @@
+        slist => { "admin@acme.com", "oncall@acme.com" },
+        if => strcmp( $(env_classification.environment), "production" );
+
+      "admin_contact"
+        slist => { "dev@acme.com" },
+        if => strcmp( $(env_classification.environment), "staging" );
 
+      "admin_contact"
+        slist => { "root@localhost" },
+        if => strcmp( $(env_classification.environment), "unknown" );
+
+      ## This is to extract the available package updates status
+      "updates_available"
+        data => packageupdatesmatching(".*", ".*", ".*", ".*");
+      "count_updates" int => length("updates_available");
+
+  classes:
+
+      # We define a class indicating there are updates available, it might be
+      # useful for various different policies.
+
+      "have_updates_available"
+        expression => isgreaterthan( $(count_updates), 0),
+        scope => "namespace";
+
+}
 bundle agent motd {
 
   vars:
-    "motd_path" string => "/etc/motd";
-    "template_path" string => "$(this.promise_filename).mustache";
-
-    ## This is to extract the cf-engine role ( hub or client )
-    "cf_role"
-      string => ifelse( "policy_server", "Policy Server", "Policy Client");
-
-    ## This is to extract the available package updates status
-    "updates_available"
-      data => packageupdatesmatching(".*", ".*", ".*", ".*");
-    "amount_updates" int => length("updates_available");
-
-    ## This is to define a production / stage example role for the server
-    "environment_list" slist => { "/etc/prod", "/etc/staging" };
-    "environment"
-      string => ifelse( filesexist( @(environment_list) ), "unknown",
-                        fileexists("/etc/prod"), "production",
-                        fileexists("/etc/staging"), "staging",
-                        "unknown" );
-
-    ## Based on the environment we define different contatcs.
-    "admin_contact"
-      slist => { "admin@acme.com", "oncall@acme.com" },
-      if => strcmp( $(environment), "production" );
-
-    "admin_contact"
-      slist => { "dev@acme.com" },
-      if => strcmp( $(environment), "staging" );
-
-    "admin_contact"
-      slist => { "root@localhost" },
-      if => strcmp( $(environment), "unknown" );
-
-  ## We define a class to conditionally render the role status in the final
-  ## motd.
-  classes:
-    "env_$(environment)" expression => "any";
+      "motd_path" string => "/tmp/motd";
+
+      # It's considered best practice to prepare a data container holding the
+      # information you need to render the template instead of relying on
+      # current datastate()
 
-  ## In order to get package inventory, we need to have a legacy package
-  ## promise.
-  packages:
-    "vim" -> { "https://dev.cfengine.com/issues/7864" }
-      package_policy => "add";
+      # First we extract currently defined classes from datastate(), then we
+      # construct the template data.
+
+      "_state" data => datastate(), if => not( isvariable ( $(this.promiser) ) );
```

Comment:
What's the purpose of the `if => not(isvariable($(this.promiser)))`? I think it needs an explanation.

Reply:
ack, added explination.